File: Python/Circular/Circular.py
class CircularNode:
    def __init__(self, data):
        self.data = data
        self.next = None

class CircularLinkedList:
    def __init__(self):
        self.head = None

    def insert(self, data):
        new_node = CircularNode(data)
        if self.head is None:
            self.head = new_node
            new_node.next = self.head
        else:
            temp = self.head
            while temp.next != self.head:
                temp = temp.next
            temp.next = new_node
            new_node.next = self.head

    def delete_node(self, key):
        if self.head is None:
            return

        current = self.head
        prev = None

        # If the node to be deleted is the head node
        if current.data == key:
            while current.next != self.head:
                current = current.next

            if self.head.next == self.head:
                self.head = None
                return

            current.next = self.head.next
            self.head = current.next
            return

        # Search for the node to be deleted
        while current.next != self.head and current.data != key:
            prev = current
            current = current.next

        if current.data == key:
            prev.next = current.next

    def display(self):
        if self.head is None:
            return
        current = self.head
        while True:
            print(current.data, end=" -> ")
            current = current.next
            if current == self.head:
                break
        print("(head)")

# Example usage
list = CircularLinkedList()
list.insert(3)
list.insert(2)
list.insert(1)

print("Original List:")
list.display()

list.delete_node(2)
print("After Deleting 2:")
list.display()

File: Java/Circular/Circular.java
class CircularNode {
    int data;
    CircularNode next;

    CircularNode(int data) {
        this.data = data;
        this.next = null;
    }
}

class CircularLinkedList {
    CircularNode head;

    public CircularLinkedList() {
        this.head = null;
    }

    public void insert(int data) {
        CircularNode newNode = new CircularNode(data);
        if (head == null) {
            head = newNode;
            newNode.next = head;
        } else {
            CircularNode temp = head;
            while (temp.next != head) {
                temp = temp.next;
            }
            temp.next = newNode;
            newNode.next = head;
        }
    }

    public void deleteNode(int key) {
        if (head == null) return;

        CircularNode current = head, prev = null;

        // If the node to be deleted is the head node
        if (current.data == key) {
            while (current.next != head) {
                current = current.next;
            }

            if (head.next == head) {
                head = null;
                return;
            }

            current.next = head.next;
            head = current.next;
            return;
        }

        // Search for the node to be deleted
        do {
            prev = current;
            current = current.next;
        } while (current != head && current.data != key);

        if (current.data == key) {
            prev.next = current.next;
        }
    }

    public void display() {
        if (head == null) return;
        CircularNode current = head;
        do {
            System.out.print(current.data + " -> ");
            current = current.next;
        } while (current != head);
        System.out.println("(head)");
    }

    public static void main(String[] args) {
        CircularLinkedList list = new CircularLinkedList();
        list.insert(3);
        list.insert(2);
        list.insert(1);

        System.out.println("Original List:");
        list.display();

        list.deleteNode(2);
        System.out.println("After Deleting 2:");
        list.display();
    }
}
